lib/src/core/models/functions/functions.dart

```diff
@@ -0,0 +1,120 @@
+import 'dart:convert';
+
+import 'package:meta/meta.dart';
+
+@immutable
+class OpenAiFunctionModel {
+  final String name;
+  final String? description;
+  final OpenAiFunctionParameters? parameters;
+
+  const OpenAiFunctionModel({
+    required this.name,
+    this.description,
+    this.parameters,
+  });
+
+  Map<String, dynamic> toMap() {
+    return {
+      'name': name,
+      if (description != null) 'description': description,
+      if (parameters != null) 'parameters': parameters!.toMap(),
+    };
+  }
+}
+
+class OpenAiFunctionParameters {
+  final String type;
+  final List<OpenAiFunctionProperty> properties;
+  OpenAiFunctionParameters({
+    this.type = 'object',
+    required this.properties,
+  });
+
+  Map<String, dynamic> toMap() {
+    final requiredProperties = properties
+        .where((property) => property.isRequired)
+        .map((property) => property.name)
+        .toList();
+    return {
+      'type': type,
+      'properties': {
+        for (final parameter in properties)
+          parameter.name: {
+            if (parameter.type != null) 'type': parameter.type,
+            if (parameter.description != null)
+              'description': parameter.description,
+            if (parameter.enumValues != null) 'enum': parameter.enumValues,
+          },
+      },
+      'required': requiredProperties,
+    };
+  }
+}
+
+class OpenAiFunctionProperty {
+  static const functionTypeInteger = 'integer';
+  static const functionTypeString = 'string';
+  static const functionTypeBoolean = 'boolean';
+  static const functionTypeFloat = 'float';
+
+  final String name;
+  final String? type;
+  final String? description;
+  final List<String>? enumValues;
+  final bool isRequired;
+
+  const OpenAiFunctionProperty({
+    required this.name,
+    this.type,
+    this.description,
+    this.isRequired = false,
+    this.enumValues,
+  });
+
+  @override
+  int get hashCode {
+    return name.hashCode ^
+        type.hashCode ^
+        description.hashCode ^
+        enumValues.hashCode;
+  }
+}
+
+@immutable
+class FunctionCall {
+  static const none = FunctionCall._(value: 'none');
```

Comment:
this is just a question, can I know why you did not used factory constructors for `auto` and `none` as well ?

Reply:
No particular reason, since they are going to be the same forever a const felt better, but it's not a big deal either way.